katello-attach-subscription

```diff
@@ -99,24 +135,769 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+if @options[:virtwho]
+  if @options[:debug]
+    puts " DEBUG: use of virt-who enabled, json module added"
+  end
+  require 'json'
+end
+
+# if the script run with check density option enabled it's required CSV to write the state of the cluster
+if @options[:density]
+  if @options[:debug]
+    puts " DEBUG: check cluster density option enabled, csv module added"
+  end
+  require 'csv'
+end
+
 # missing configuration parameters from file or command line are taken from the default one
 @defaults.each do |key,val|
   if not @options.has_key?(key)
     @options[key] = val
   end
 end
 
+@virtwho_data = false
+@parsed_hypervisors_array = []
+@full_cluster  = []
+@empty_cluster = []
+@clustered_hypervisor_list = []
+
+# check functions
+def checksubs()
+  # initialize variables
+  # by default any kind of subscription is fine
+  subfiltertype = nil
+  cachefile = @options[:cachefile].to_s + "_org" + @options[:org].to_s
+  cachepresent = false
+  if @options[:usecache]
+    # if the file exists, let's open it
+    if File.file?(cachefile)
+      begin
+        cacheyaml = YAML.load_file(cachefile)
+        if @yaml.has_key?(:subs) and @yaml[:subs].is_a?(Hash) and @yaml[:subs].count > 0
+          @yaml[:subs] = cacheyaml['subs']
+          puts "Reading subscription from cache"
+          cachepresent = true
+        else
+          puts "Empty subscription cache, continuing"
+        end
+      rescue Exception
+        abort "FATAL ERROR: Failed to read cache file for subscriptions. re-run without --read-from-cache."
+      end
+    else
+      puts "FATAL ERROR: --read-from-cache option requested, but the cache file is not currently present."
+      exit 5
+    end
+  end
+
+  # create api binding
+  api = ApipieBindings::API.new({:uri => @options[:uri], :username => @options[:user], :password => @options[:pass], :api_version => '2', :timeout => @options[:timeout]})
+
+  # sanity check and fix of yaml file for subscription
+  # going to check if any id in the yaml do not exists, and expand the search keys
+  # sub will keep the originally requested subscriptions, sub_parsed the found one
+  if not cachepresent
+    puts "Subscription parsing started. Please be patient."
+    subsyamltotalentry = @yaml[:subs].count
+
+    @yaml[:subs].each_with_index do |sub, subcurrentcount|
+      puts " Started parsing entry #{subcurrentcount+1}/#{subsyamltotalentry}"
+      # check if the hash contains data. If not, skip it.
+      if not sub.is_a?(Hash)
+        puts "  YAML entry without valid data, like empty entry. No further actions on this entry."
+        next
+      end
+
+      # starting the process
+      if @options[:verbose]
+        puts " VERBOSE: started parsing for entry '#{sub["hostname"]}'"
+      end
+
+      # If type is Hypervisor, only NORMAL subscription has to be considered
+      if sub["type"] == "Hypervisor"
+        subfiltertype = "NORMAL"
```

Comment:
`subfiltertype` is not initialized to `nil` on each run of this `each_with_index`, so after you've set it once to `NORMAL` it will never be reset.